line comments

dm0n3y opened the pull request.

YuanhaoMeng committed:
Add LineComment variant

dm0n3y committed:
Refactor UHExp

Commit:
Merge branch 'dev' into line-comments

dm0n3y committed:
wip

YuanhaoMeng committed:
ZExp.re add CommentLine

YuanhaoMeng committed:
Adding CommentLine in CursorPath.re

YuanhaoMeng committed:
DocOfTerm.re, Action.re, Dynamics.re, Statics.re add CommentLine    -----NOTE: Action.re is not finished

YuanhaoMeng committed:
Action.re add CommentLine (line 3762 & 3777 not finished)

YuanhaoMeng committed:
Adding CommentLine to Action.re

YuanhaoMeng committed:
Adding CommentLine to Action.re

YuanhaoMeng committed:
fix commentline in Action.re

dm0n3y committed:
Kick off Code.re refactor

YuanhaoMeng committed:
add CommentLine in Code.re

YuanhaoMeng committed:
Please enter the commit message for your changes. Lines starting

YuanhaoMeng committed:
Adding commentline latest

YuanhaoMeng committed:
Merge remote-tracking branch 'origin/dev' into line-comments

YuanhaoMeng committed:
Merge remote-tracking branch 'origin/dev' into line-comments

YuanhaoMeng committed:
commentline(break)

YuanhaoMeng committed:
Merge remote-tracking branch 'origin/dev' into line-comments

dm0n3y committed:
Fix bug, add comments

YuanhaoMeng committed:
Merge branch 'line-comments' of https://github.com/hazelgrove/hazel into line-comments

YuanhaoMeng committed:
Adding CommentLine

YuanhaoMeng committed:
CommentLine Added

YuanhaoMeng committed:
CommentLine

YuanhaoMeng committed:
Some modifications

YuanhaoMeng committed:
delete debug.log file

YuanhaoMeng committed:
Add basic SubCommentLine feature.

YuanhaoMeng committed:
enable multi-line comments, with some bugs

YuanhaoMeng committed:
change the indicator for subcommentline from $ to |

YuanhaoMeng committed:
multi-comment lines

dm0n3y commented:
Opening PR so that we have a place to view and discuss incoming merge

dm0n3y committed:
Merge remote-tracking branch 'origin/dev' into line-comments

dm0n3y committed:
Fix broken comment construction

dm0n3y commented:
I think I've reached parity with your changes so far (which you can view [here](https://github.com/hazelgrove/hazel/compare/dev...833734957a158578ae8712cf607b2e01810e2b87)). Comments work like before, but there's no styling applied. I'll leave that to you as an exercise by which you can get used to new view code (ask me questions or set up a time to pair with me).

YuanhaoMeng committed:
fix some bugs

dm0n3y committed:
Merge branch 'dev' into line-comments

YuanhaoMeng committed:
Disable SubCommentLine function

YuanhaoMeng committed:
delete SubCommentLine feature

YuanhaoMeng committed:
with 2 bugs

YuanhaoMeng committed:
CommentLine full

cyrus- commented:
Hey Yuanhao, you should merge the latest dev changes into your branch next.

dm0n3y committed:
Merge remote-tracking branch 'origin/dev' into line-comments

dm0n3y committed:
Add comment line styling

dm0n3y committed:
Add docs

dm0n3y committed:
Enable comment lines in block conclusion

dm0n3y committed:
Remove unnecessary styling

dm0n3y commented on src/hazelcore/semantics/ZExp.re:
```diff
@@ -57,6 +65,14 @@ let line_can_be_swapped = (line: zline): bool =>
   switch (line) {
+  | CommentLine(comment) =>
+    // f(x) == x |> f
+    // [1, 2] @ [3, 4] == [1, 2, 3, 4]
```
Remove

dm0n3y commented on src/hazelcore/semantics/ZExp.re:
```diff
@@ -5,6 +5,14 @@ type t = zblock
+  // CursorL(OnDelim(0, After), CommentLine("goodbye"))
+  // #| goodbye
+  // CursorL(OnText(0), CommentLine("goodbye"))
+  // # |goodbye
```
Remove

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2483,6 +2499,41 @@ module Exp = {
+        let new_ze = (prefix, new_zline, new_suffix);
+        Succeeded(SynDone((new_ze, ty, u_gen)));
+      | _ => Failed
+      }
```
Rather than repeating the pattern matching on the prefix/suffix and zline (both in the `when` guard and in the rule clause), I would refactor the top level switch expression of `syn_perform_block` so that you pattern match on `(a, (ListUtil.split_last(prefix), zline, suffix))`. This will allow you to do the pattern matching just once.

dm0n3y replied:
You'll need to repeat these cases in `ana_perform_block` too.

dm0n3y replied:
Still need to address these comments

dm0n3y reviewed (commented):
@YuanhaoMeng Looks good overall! Requested a few changes in comments, should be good to merge once those are addressed.

dm0n3y committed:
Merge remote-tracking branch 'origin/dev' into line-comments

cyrus- commented:
@YuanhaoMeng this is 96 commits behind dev, and there are some comments David left above that you need to address before we can merge

cyrus- commented:
@YuanhaoMeng you marked this as ready for review without making any changes since my last comments. please make those changes. contact me if you're confused.

YuanhaoMeng committed:
delete some comments

YuanhaoMeng committed:
Merge branch 'line-comments' of https://github.com/hazelgrove/hazel into line-comments

YuanhaoMeng commented:
I am sorry that I misunderstood the "comment", I thought i was supposed to modify some code layout.
Now those comments are removed.

cyrus- commented:
@YuanhaoMeng your branch is behind `dev` and has conflicts with `dev`. Please merge the latest `dev` changes into your branch before marking it ready for review.

dm0n3y commented:
@YuanhaoMeng https://github.com/hazelgrove/hazel/pull/94#discussion_r412540043

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Fix from merge

Accidentally recreated files

nj-wilson committed:
line-comments catchup

Various small additions for switches

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Fix mk_line for CommentLine

Missed this from the merge

dm0n3y commented on src/hazelweb/gui/ActionPanel.re:
```diff
@@ -482,6 +482,7 @@ let _check_actions = (a: Action_common.t) =>
   | Construct(SAsc) => Added
   | Construct(SOp(SEquals)) => Added
   | Construct(SLine) => Added
+  | Construct(SCommentLine) => Added
```
Not seeing an actual action panel entry for this

dm0n3y commented on src/hazelweb/gui/UndoHistoryPanel.re:
```diff
@@ -258,6 +258,7 @@ let view = (~inject: ModelAction.t => Vdom.Event.t, model: Model.t) => {
     | Line(_, line_content) =>
       switch (line_content) {
       | EmptyLine => indicate_words_view("empty line")
+      | CommentLine(_) => indicate_words_view("comment")
```
@PPER any thoughts here on how comment lines should appear in undo panel? Here is what it looks like now
<img width="1638" alt="image" src="https://user-images.githubusercontent.com/8718124/87466208-df21be00-c5e3-11ea-9dc6-9249c409e05c.png">

dm0n3y commented on src/hazelcore/ZExp.re:
```diff
@@ -175,6 +182,39 @@ and is_before_zoperand =
   | ApPaletteZ(_) => false;
+
+// The following 2 functions are specifically for CommentLines!!
+// Check if the cursor at "OnDelim(After)" in a (sub) "CommentLine"
```
I don't think sub or pre comment lines are a thing anymore, let's remove those qualifiers

dm0n3y reviewed (commented):
Thanks for getting this up to speed! Couple requests w.r.t. new things (e.g. action panel, undo panel) that weren't around when this work started.

nj-wilson committed:
Remove references to sub, pre comments

nj-wilson committed:
Add creating comment line to Action panel

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Fix from merge

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

nj-wilson committed:
Ignore spaces when inside a comment line

nj-wilson committed:
Return correct length of comment

nj-wilson committed:
History panel shows new comment as "empty comment"

nj-wilson committed:
Differentiate spaces in comments from everywhere else

Emits a different Construct based on if a space is detected within a comment line or not. Simplifies comment line code, supersedes previous fix

nj-wilson committed:
Remove leftover

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

dm0n3y reviewed (commented):
Seems I can't insert anything on an empty line

nj-wilson committed:
Fix typing on blank lines

nj-wilson committed:
Merge remote-tracking branch 'origin/dev' into line-comments

dm0n3y commented:
Looks good. Thanks @nj-wilson!
